pkg/dynamicaccess/accesslogic.go

```diff
@@ -0,0 +1,156 @@
+	// Encrypts a Swarm reference for a given grantee
+	EncryptRef(ctx context.Context, storage kvs.KeyValueStore, grantee *ecdsa.PublicKey, ref swarm.Address) (swarm.Address, error)
+}
+
+type ActLogic struct {
+	Session
+}
+
+var _ Control = (*ActLogic)(nil)
+
+// Adds a new publisher to an empty act
+func (al ActLogic) AddPublisher(ctx context.Context, storage kvs.KeyValueStore, publisher *ecdsa.PublicKey) error {
+	accessKey := encryption.GenerateRandomKey(encryption.KeyLength)
+
+	return al.AddGrantee(ctx, storage, publisher, publisher, &accessKey)
+}
+
+// Encrypts a SWARM reference for a publisher
+func (al ActLogic) EncryptRef(ctx context.Context, storage kvs.KeyValueStore, publisherPubKey *ecdsa.PublicKey, ref swarm.Address) (swarm.Address, error) {
+	accessKey, err := al.getAccessKey(ctx, storage, publisherPubKey)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	refCipher := encryption.New(accessKey, 0, uint32(0), hashFunc)
+	encryptedRef, _ := refCipher.Encrypt(ref.Bytes())
+
+	return swarm.NewAddress(encryptedRef), nil
+}
+
+// Adds a new grantee to the ACT
+func (al ActLogic) AddGrantee(ctx context.Context, storage kvs.KeyValueStore, publisherPubKey, granteePubKey *ecdsa.PublicKey, accessKeyPointer *encryption.Key) error {
+	var accessKey encryption.Key
+	var err error // Declare the "err" variable
+
+	if accessKeyPointer == nil {
+		// Get previously generated access key
+		accessKey, err = al.getAccessKey(ctx, storage, publisherPubKey)
+		if err != nil {
+			return err
+		}
+	} else {
+		// This is a newly created access key, because grantee is publisher (they are the same)
+		accessKey = *accessKeyPointer
+	}
+
+	// Encrypt the access key for the new Grantee
+	keys, err := al.getKeys(granteePubKey)
+	if err != nil {
+		return err
+	}
+	lookupKey := keys[0]
+	// accessKeyDecryptionKey is used for encryption of the access key
+	accessKeyDecryptionKey := keys[1]
+
+	// Encrypt the access key for the new Grantee
+	cipher := encryption.New(encryption.Key(accessKeyDecryptionKey), 0, uint32(0), hashFunc)
+	granteeEncryptedAccessKey, err := cipher.Encrypt(accessKey)
+	if err != nil {
+		return err
+	}
+
+	// Add the new encrypted access key for the Act
+	return storage.Put(ctx, lookupKey, granteeEncryptedAccessKey)
+}
+
+// Will return the access key for a publisher (public key)
+func (al *ActLogic) getAccessKey(ctx context.Context, storage kvs.KeyValueStore, publisherPubKey *ecdsa.PublicKey) ([]byte, error) {
+	keys, err := al.getKeys(publisherPubKey)
+	if err != nil {
+		return nil, err
+	}
+	publisherLookupKey := keys[0]
+	publisherAKDecryptionKey := keys[1]
+	// no need to constructor call if value not found in act
+	accessKeyDecryptionCipher := encryption.New(encryption.Key(publisherAKDecryptionKey), 0, uint32(0), hashFunc)
+	encryptedAK, err := storage.Get(ctx, publisherLookupKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return accessKeyDecryptionCipher.Decrypt(encryptedAK)
+}
+
+var oneByteArray = []byte{1}
```

Comment:
do you need these anywhere else than within `getKeys` function? move it inside if not

Reply:
oneByteArray is reused now in the controller for grantee list encryption